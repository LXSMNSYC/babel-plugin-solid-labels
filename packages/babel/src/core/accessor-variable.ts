import type * as babel from '@babel/core';
import * as t from '@babel/types';
import derefMemo from './deref-memo';

export default function accessorVariable(
  path: babel.NodePath,
  accessorIdentifier: t.Identifier,
  callee: t.Identifier,
  replacement: Array<t.Expression | t.SpreadElement | t.JSXNamespacedName | t.ArgumentPlaceholder>,
): t.VariableDeclarator {
  const readIdentifier = path.scope.generateUidIdentifier(accessorIdentifier.name);

  derefMemo(
    path,
    accessorIdentifier,
    readIdentifier,
  );

  return t.variableDeclarator(
    readIdentifier,
    t.callExpression(callee, replacement),
  );
}
