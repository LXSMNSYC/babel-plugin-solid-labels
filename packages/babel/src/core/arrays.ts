export function forEach<T>(arr: T[], callback: (value: T, index: number) => (boolean | void)) {
  for (let i = 0, len = arr.length; i < len; i += 1) {
    if (callback(arr[i], i)) {
      break;
    }
  }
}

export function some<T>(arr: T[], callback: (value: T, index: number) => boolean) {
  for (let i = 0, len = arr.length; i < len; i += 1) {
    if (callback(arr[i], i)) {
      return true;
    }
  }
  return false;
}

export function every<T>(arr: T[], callback: (value: T, index: number) => boolean) {
  for (let i = 0, len = arr.length; i < len; i += 1) {
    if (!callback(arr[i], i)) {
      return false;
    }
  }
  return true;
}
