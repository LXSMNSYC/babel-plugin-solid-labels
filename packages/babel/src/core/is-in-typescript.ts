import * as t from '@babel/types';

export default function isInTypeScript(path: babel.NodePath): boolean {
  let parent = path.parentPath;
  while (parent) {
    if (t.isTypeScript(parent.node)) {
      return true;
    }
    parent = parent.parentPath;
  }
  return false;
}
