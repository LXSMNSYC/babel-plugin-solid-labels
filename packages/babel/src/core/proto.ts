import type * as babel from '@babel/core';
import * as t from '@babel/types';

interface ProtoObjectState {
  hasProto: boolean;
  root: t.ObjectExpression;
  proto: t.ObjectExpression;
  getters?: t.ObjectExpression;
  setters?: t.ObjectExpression;
}

const ROOT_GET = 'get';
const ROOT_SET = 'set';
const ROOT_SYMBOL = '___'; // TODO

function generateProto(
  path: babel.NodePath<t.ObjectExpression>,
): void {
  const current = (path.state || {}) as ProtoObjectState;
  if (!current.hasProto) {
    const protoID = path.scope.generateUidIdentifier('proto');
    const proto = t.objectExpression([]);
    path.scope.push({
      id: protoID,
      init: proto,
      kind: 'const',
    });
    current.proto = proto;

    const root = t.objectExpression([]);
    path.node.properties.push(
      t.objectProperty(t.identifier(ROOT_SYMBOL), root),
      t.objectProperty(t.identifier('__proto__'), protoID),
    );
    current.root = root;
    current.hasProto = true;
  }
  path.state = current;
}

function initProtoGetters(
  path: babel.NodePath<t.ObjectExpression>,
  property: t.ObjectProperty,
): void {
  generateProto(path);
  const current = path.state as ProtoObjectState;
  if (current.getters) {
    current.getters.properties.push(property);
  } else {
    const rootGet = t.objectExpression([property]);
    current.root.properties.push(
      t.objectProperty(t.identifier(ROOT_GET), rootGet),
    );
    current.getters = rootGet;
  }
}

function registerProtoGetter(
  path: babel.NodePath<t.ObjectExpression>,
  identifier: t.Identifier,
): void {
  const targetProperty = t.memberExpression(
    t.memberExpression(
      t.memberExpression(t.identifier('this'), t.identifier(ROOT_SYMBOL)),
      t.identifier(ROOT_GET),
    ),
    identifier,
  );
  const current = path.state as ProtoObjectState;
  current.proto.properties.push(
    t.objectMethod('get', identifier, [], t.blockStatement([
      t.returnStatement(
        t.callExpression(targetProperty, []),
      ),
    ])),
  );
}

function addUnoptimizedGetter(
  property: babel.NodePath<t.ObjectProperty>,
  source: t.Identifier,
): void {
  if (!t.isPrivateName(property.node.key)) {
    property.replaceWith(
      t.objectMethod(
        'get',
        property.node.key,
        [],
        t.blockStatement([
          t.returnStatement(
            t.callExpression(
              source,
              [],
            ),
          ),
        ]),
        property.node.computed,
      ),
    );
  }
}

export function addProtoGetter(
  path: babel.NodePath<t.ObjectExpression>,
  property: babel.NodePath<t.ObjectProperty>,
  source: t.Identifier,
): void {
  if (t.isIdentifier(property.node.key)) {
    const identifier = property.node.key;
    initProtoGetters(path, t.objectProperty(identifier, source));
    registerProtoGetter(path, identifier);
    property.remove();
  } else {
    addUnoptimizedGetter(property, source);
  }
}

function initProtoSetters(
  path: babel.NodePath<t.ObjectExpression>,
  property: t.ObjectProperty,
): void {
  generateProto(path);
  const current = path.state as ProtoObjectState;
  if (current.setters) {
    current.setters.properties.push(property);
  } else {
    const rootSet = t.objectExpression([property]);
    current.root.properties.push(
      t.objectProperty(t.identifier(ROOT_SET), rootSet),
    );
    current.setters = rootSet;
  }
}

function registerProtoSetter(
  path: babel.NodePath<t.ObjectExpression>,
  identifier: t.Identifier,
): void {
  const targetProperty = t.memberExpression(
    t.memberExpression(
      t.memberExpression(t.identifier('this'), t.identifier(ROOT_SYMBOL)),
      t.identifier(ROOT_SET),
    ),
    identifier,
  );
  const current = path.state as ProtoObjectState;
  const param = path.scope.generateUidIdentifier('param');
  current.proto.properties.push(
    t.objectMethod('set', identifier, [param], t.blockStatement([
      t.expressionStatement(
        t.callExpression(targetProperty, [
          t.arrowFunctionExpression([], param),
        ]),
      ),
    ])),
  );
}

function addUnoptimizedSetter(
  property: babel.NodePath<t.ObjectProperty>,
  source: t.Identifier,
): void {
  if (!t.isPrivateName(property.node.key)) {
    const param = property.scope.generateUidIdentifier('param');
    property.replaceWith(
      t.objectMethod(
        'set',
        property.node.key,
        [param],
        t.blockStatement([
          t.expressionStatement(
            t.callExpression(
              source,
              [t.arrowFunctionExpression([], param)],
            ),
          ),
        ]),
        property.node.computed,
      ),
    );
  }
}

export function addProtoSetter(
  path: babel.NodePath<t.ObjectExpression>,
  property: babel.NodePath<t.ObjectProperty>,
  source: t.Identifier,
): void {
  if (t.isIdentifier(property.node.key)) {
    const identifier = property.node.key;
    initProtoSetters(path, t.objectProperty(identifier, source));
    registerProtoSetter(path, identifier);
    property.remove();
  } else {
    addUnoptimizedSetter(property, source);
  }
}

export function addProtoProperty(
  path: babel.NodePath<t.ObjectExpression>,
  property: babel.NodePath<t.ObjectProperty>,
  readSource: t.Identifier,
  writeSource: t.Identifier,
): void {
  if (t.isIdentifier(property.node.key)) {
    const identifier = property.node.key;
    initProtoGetters(path, t.objectProperty(identifier, readSource));
    initProtoSetters(path, t.objectProperty(identifier, writeSource));
    registerProtoGetter(path, identifier);
    registerProtoSetter(path, identifier);
    property.remove();
  } else {
    addUnoptimizedGetter(property, readSource);
    addUnoptimizedSetter(property, writeSource);
  }
}
