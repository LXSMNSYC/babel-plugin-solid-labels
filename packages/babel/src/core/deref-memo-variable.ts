import * as babel from '@babel/core';
import * as t from '@babel/types';
import derefMemo from './deref-memo';

export default function derefMemoVariable(
  path: babel.NodePath,
  memoIdentifier: t.Identifier,
  stateIdentifier: t.Expression,
) {
  const readIdentifier = path.scope.generateUidIdentifier(memoIdentifier.name);

  derefMemo(
    path,
    memoIdentifier,
    readIdentifier,
  );

  return t.variableDeclarator(readIdentifier, stateIdentifier);
}
