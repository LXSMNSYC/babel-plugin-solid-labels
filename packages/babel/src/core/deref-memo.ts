import * as babel from '@babel/core';
import * as t from '@babel/types';
import { unexpectedType } from './errors';
import isInTypeScript from './is-in-typescript';
import unwrapNode from './unwrap-node';

const REF_MEMO_CTF = '$refMemo';
const GET_CTF = '$get';

const CALL_CTF = new Set([REF_MEMO_CTF, GET_CTF]);

const GETTER_CTF = '$getter';
const PROPERTY_CTF = '$property';

const OBJECT_PROPERTY_CTF = new Set([GETTER_CTF, PROPERTY_CTF]);

export default function derefMemo(
  path: babel.NodePath,
  memoIdentifier: t.Identifier,
  readIdentifier: t.Identifier,
): void {
  path.scope.path.traverse({
    CallExpression(p) {
      if (p.scope !== path.scope && p.scope.hasOwnBinding(memoIdentifier.name)) {
        return;
      }
      const trueCallee = unwrapNode(p.node.callee, t.isIdentifier);
      if (!trueCallee || p.scope.hasBinding(trueCallee.name) || !CALL_CTF.has(trueCallee.name)) {
        return;
      }
      const arg = unwrapNode(p.node.arguments[0], t.isIdentifier);
      if (!arg) {
        throw unexpectedType(p, p.node.arguments[0].type, 'Identifier');
      }
      if (arg.name !== memoIdentifier.name) {
        return;
      }
      if (trueCallee.name === REF_MEMO_CTF) {
        p.replaceWith(readIdentifier);
      }
      if (trueCallee.name === GET_CTF) {
        p.replaceWith(readIdentifier);
      }
    },
    ObjectProperty(p) {
      if (p.scope !== path.scope && p.scope.hasOwnBinding(memoIdentifier.name)) {
        return;
      }
      if (p.node.shorthand && !p.node.computed) {
        if (
          t.isIdentifier(p.node.key)
          && t.isIdentifier(p.node.value)
          && p.node.key.name === memoIdentifier.name
          && p.node.value.name === memoIdentifier.name
        ) {
          p.replaceWith(
            t.objectProperty(
              p.node.key,
              t.callExpression(readIdentifier, []),
            ),
          );
        }
        return;
      }
      const trueCallExpr = unwrapNode(p.node.value, t.isCallExpression);
      if (trueCallExpr) {
        const trueCallee = unwrapNode(trueCallExpr.callee, t.isIdentifier);
        if (
          !trueCallee
          || p.scope.hasBinding(trueCallee.name)
          || !OBJECT_PROPERTY_CTF.has(trueCallee.name)
        ) {
          return;
        }
        if (t.isPrivateName(p.node.key)) {
          throw unexpectedType(p, 'PrivateName', 'Expression');
        }
        const arg = trueCallExpr.arguments[0];
        if (!t.isIdentifier(arg)) {
          throw unexpectedType(p, arg.type, 'Identifier');
        }
        if (arg.name !== memoIdentifier.name) {
          return;
        }
        if (trueCallee.name === GETTER_CTF) {
          p.replaceWith(
            t.objectMethod(
              'get',
              p.node.key,
              [],
              t.blockStatement([
                t.returnStatement(
                  t.callExpression(
                    readIdentifier,
                    [],
                  ),
                ),
              ]),
            ),
          );
        }
        if (trueCallee.name === PROPERTY_CTF) {
          p.replaceWith(
            t.objectMethod(
              'get',
              p.node.key,
              [],
              t.blockStatement([
                t.returnStatement(
                  t.callExpression(
                    readIdentifier,
                    [],
                  ),
                ),
              ]),
            ),
          );
        }
      }
    },
    Expression(p) {
      if (p.scope !== path.scope && p.scope.hasOwnBinding(memoIdentifier.name)) {
        return;
      }
      if (
        t.isIdentifier(p.node)
        && !isInTypeScript(p)
        && p.node.name === memoIdentifier.name
      ) {
        p.replaceWith(t.callExpression(readIdentifier, []));
      }
    },
  });
}
