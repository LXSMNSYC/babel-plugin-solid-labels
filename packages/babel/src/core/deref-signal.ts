import * as babel from '@babel/core';
import * as t from '@babel/types';
import assert from './assert';
import { unexpectedType } from './errors';
import isAwaited from './is-awaited';
import isYielded from './is-yielded';
// import isInTypeScript from './is-in-typescript';
import unwrapNode from './unwrap-node';

const REF_SIGNAL_CTF = '$refSignal';
const GET_CTF = '$get';
const SET_CTF = '$set';

const CALL_CTF = new Set([REF_SIGNAL_CTF, GET_CTF, SET_CTF]);

const GETTER_CTF = '$getter';
const SETTER_CTF = '$setter';
const PROPERTY_CTF = '$property';

const OBJECT_PROPERTY_CTF = new Set([GETTER_CTF, SETTER_CTF, PROPERTY_CTF]);

export default function derefSignal(
  path: babel.NodePath,
  signalIdentifier: t.Identifier,
  readIdentifier: t.Identifier,
  writeIdentifier: t.Identifier,
): void {
  path.scope.path.traverse({
    CallExpression(p) {
      if (p.scope !== path.scope && p.scope.hasOwnBinding(signalIdentifier.name)) {
        return;
      }
      const trueCallee = unwrapNode(p.node.callee, t.isIdentifier);
      if (!trueCallee || p.scope.hasBinding(trueCallee.name) || !CALL_CTF.has(trueCallee.name)) {
        return;
      }
      const rawArgs = p.node.arguments[0];
      const arg = unwrapNode(rawArgs, t.isIdentifier);
      assert(arg, unexpectedType(p, rawArgs.type, 'Identifier'));
      if (arg.name !== signalIdentifier.name) {
        return;
      }
      if (trueCallee.name === REF_SIGNAL_CTF) {
        p.replaceWith(
          t.arrayExpression([
            readIdentifier,
            writeIdentifier,
          ]),
        );
      }
      if (trueCallee.name === GET_CTF) {
        p.replaceWith(readIdentifier);
      }
      if (trueCallee.name === SET_CTF) {
        p.replaceWith(writeIdentifier);
      }
    },
    ObjectProperty(p) {
      if (p.scope !== path.scope && p.scope.hasOwnBinding(signalIdentifier.name)) {
        return;
      }
      if (p.node.shorthand && !p.node.computed) {
        if (
          t.isIdentifier(p.node.key)
          && t.isIdentifier(p.node.value)
          && p.node.key.name === signalIdentifier.name
          && p.node.value.name === signalIdentifier.name
        ) {
          p.replaceWith(
            t.objectProperty(
              p.node.key,
              t.callExpression(readIdentifier, []),
            ),
          );
        }
        return;
      }
      const trueCallExpr = unwrapNode(p.node.value, t.isCallExpression);
      if (trueCallExpr) {
        const trueCallee = unwrapNode(trueCallExpr.callee, t.isIdentifier);
        if (
          !trueCallee
          || p.scope.hasBinding(trueCallee.name)
          || !OBJECT_PROPERTY_CTF.has(trueCallee.name)
        ) {
          return;
        }
        assert(!t.isPrivateName(p.node.key), unexpectedType(p, 'PrivateName', 'Expression'));
        const arg = trueCallExpr.arguments[0];
        assert(t.isIdentifier(arg), unexpectedType(p, arg.type, 'Identifier'));
        if (arg.name !== signalIdentifier.name) {
          return;
        }
        if (trueCallee.name === SETTER_CTF) {
          const param = p.scope.generateUidIdentifier('value');
          p.replaceWith(
            t.objectMethod(
              'set',
              p.node.key,
              [param],
              t.blockStatement([
                t.returnStatement(
                  t.callExpression(
                    writeIdentifier,
                    [
                      t.arrowFunctionExpression(
                        [],
                        param,
                      ),
                    ],
                  ),
                ),
              ]),
            ),
          );
        }
        if (trueCallee.name === GETTER_CTF) {
          p.replaceWith(
            t.objectMethod(
              'get',
              p.node.key,
              [],
              t.blockStatement([
                t.returnStatement(
                  t.callExpression(
                    readIdentifier,
                    [],
                  ),
                ),
              ]),
            ),
          );
        }
        if (trueCallee.name === PROPERTY_CTF) {
          const param = p.scope.generateUidIdentifier('value');
          p.replaceWithMultiple([
            t.objectMethod(
              'get',
              p.node.key,
              [],
              t.blockStatement([
                t.returnStatement(
                  t.callExpression(
                    readIdentifier,
                    [],
                  ),
                ),
              ]),
            ),
            t.objectMethod(
              'set',
              p.node.key,
              [param],
              t.blockStatement([
                t.returnStatement(
                  t.callExpression(
                    writeIdentifier,
                    [
                      t.arrowFunctionExpression(
                        [],
                        param,
                      ),
                    ],
                  ),
                ),
              ]),
            ),
          ]);
        }
      }
    },
    Expression(p) {
      if (p.scope !== path.scope && p.scope.hasOwnBinding(signalIdentifier.name)) {
        return;
      }
      if (
        t.isIdentifier(p.node)
        // && !isInTypeScript(p)
        && p.node.name === signalIdentifier.name
      ) {
        p.replaceWith(t.callExpression(readIdentifier, []));
      }
    },
    UpdateExpression(p) {
      if (p.scope !== path.scope && p.scope.hasOwnBinding(signalIdentifier.name)) {
        return;
      }
      if (t.isIdentifier(p.node.argument) && p.node.argument.name === signalIdentifier.name) {
        const param = p.scope.generateUidIdentifier('current');
        if (p.node.prefix) {
          p.replaceWith(
            t.callExpression(
              writeIdentifier,
              [
                t.arrowFunctionExpression(
                  [param],
                  t.binaryExpression(
                    p.node.operator === '++' ? '+' : '-',
                    param,
                    t.numericLiteral(1),
                  ),
                ),
              ],
            ),
          );
        } else {
          p.replaceWith(
            t.callExpression(
              t.arrowFunctionExpression(
                [],
                t.blockStatement([
                  t.variableDeclaration(
                    'const',
                    [
                      t.variableDeclarator(
                        param,
                        t.callExpression(
                          readIdentifier,
                          [],
                        ),
                      ),
                    ],
                  ),
                  t.expressionStatement(
                    t.callExpression(
                      writeIdentifier,
                      [
                        t.arrowFunctionExpression(
                          [],
                          t.binaryExpression(
                            p.node.operator === '++' ? '+' : '-',
                            param,
                            t.numericLiteral(1),
                          ),
                        ),
                      ],
                    ),
                  ),
                  t.returnStatement(
                    param,
                  ),
                ]),
              ),
              [],
            ),
          );
        }
      }
    },
    AssignmentExpression(p) {
      if (p.scope !== path.scope && p.scope.hasOwnBinding(signalIdentifier.name)) {
        return;
      }
      const identifier = p.node.left;
      let expression = p.node.right;
      if (isAwaited(expression) || isYielded(expression)) {
        const statement = p.getStatementParent();
        const functionParent = p.getFunctionParent();
        if (statement) {
          const awaitedID = statement.scope.generateUidIdentifier('tmp');
          const declaration = t.variableDeclaration(
            'const',
            [t.variableDeclarator(
              awaitedID,
              expression,
            )],
          );

          if (functionParent) {
            if (functionParent.isAncestor(statement)) {
              statement.insertBefore(declaration);
            } else {
              functionParent.scope.push({
                id: awaitedID,
                init: expression,
                kind: 'const',
              });
            }
          } else {
            statement.insertBefore(declaration);
          }
          expression = awaitedID;
        }
      }
      if (t.isIdentifier(identifier) && identifier.name === signalIdentifier.name) {
        let arg: t.Expression;
        if (p.node.operator === '=') {
          arg = t.arrowFunctionExpression([], expression);
        } else {
          const param = p.scope.generateUidIdentifier('current');
          arg = t.arrowFunctionExpression(
            [param],
            t.assignmentExpression(
              p.node.operator,
              param,
              expression,
            ),
          );
        }
        p.replaceWith(t.callExpression(writeIdentifier, [arg]));
      }
    },
  });
}
