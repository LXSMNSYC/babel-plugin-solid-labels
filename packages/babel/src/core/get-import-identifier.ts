import * as t from '@babel/types';
import * as babel from '@babel/core';
import { addNamed } from '@babel/helper-module-imports';
import { State } from './types';

export default function getImportIdentifier(
  state: State,
  path: babel.NodePath,
  name: string,
  source: string,
): t.Identifier {
  const id = `${source}[${name}]`;
  const current = state.hooks.get(id);
  if (current) {
    return current;
  }
  const newID = addNamed(path, name, source);
  state.hooks.set(id, newID);
  return newID;
}
