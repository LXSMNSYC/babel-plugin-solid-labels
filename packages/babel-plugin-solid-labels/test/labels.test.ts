import * as babel from '@babel/core';
import plugin from '../src';
import { describe, expect, it } from 'vitest';

async function compile(code: string, dev?: boolean) {
  const result = await babel.transformAsync(code, {
    plugins: [
      [plugin, { dev }],
    ],
    parserOpts: {
      plugins: [
        'jsx',
      ],
    },
  });

  return result?.code ?? '';
}

describe('labels', () => {
  describe('signal', () => {
    it('should transform labelled variables', async () => {
      expect(await compile('signal: var count = 0')).toMatchSnapshot();
    });
    it('should transform labelled variables and add name', async () => {
      expect(await compile('signal: var count = 0', true)).toMatchSnapshot();
    });
    it('should variable read', async () => {
      expect(await compile(`signal: var count = 0; console.log(count);`)).toMatchSnapshot();
    });
    it('should variable write', async () => {
      expect(await compile(`signal: var count = 0; count = 1`)).toMatchSnapshot();
    });
  });
  describe('memo', () => {
    it('should transform labelled variables', async () => {
      expect(await compile('memo: var count = 0')).toMatchSnapshot();
    });
    it('should transform labelled variables and add name', async () => {
      expect(await compile('memo: var count = 0', true)).toMatchSnapshot();
    });
    it('should variable read', async () => {
      expect(await compile(`memo: var count = 0; console.log(count);`)).toMatchSnapshot();
    });
  });
  describe('effect', () => {
    it('should transform labels', async () => {
      expect(await compile('effect: { console.log(count); }')).toMatchSnapshot();
    });
    it('should transform labels with names', async () => {
      expect(await compile('effect: countLog: { console.log(count); }')).toMatchSnapshot();
    });
    it('should transform labels with expressions', async () => {
      expect(await compile('effect: myFunction;')).toMatchSnapshot();
    });
  });
  describe('renderEffect', () => {
    it('should transform labels', async () => {
      expect(await compile('renderEffect: { console.log(count); }')).toMatchSnapshot();
    });
    it('should transform labels with names', async () => {
      expect(await compile('renderEffect: countLog: { console.log(count); }')).toMatchSnapshot();
    });
    it('should transform labels with expressions', async () => {
      expect(await compile('renderEffect: myFunction;')).toMatchSnapshot();
    });
  });
  describe('computed', () => {
    it('should transform labels', async () => {
      expect(await compile('computed: { console.log(count); }')).toMatchSnapshot();
    });
    it('should transform labels with names', async () => {
      expect(await compile('computed: countLog: { console.log(count); }')).toMatchSnapshot();
    });
    it('should transform labels with expressions', async () => {
      expect(await compile('computed: myFunction;')).toMatchSnapshot();
    });
  });
  describe('$', () => {
    it('should transform labels with variable declarations', async () => {
      expect(await compile('$: var x = 0;')).toMatchSnapshot();
    });
    it('should transform labels with expressions', async () => {
      expect(await compile('$: console.log(x);')).toMatchSnapshot();
    });
    it('should transform labels with block expressions', async () => {
      expect(await compile('$: { console.log(x); }')).toMatchSnapshot();
    });
  });
  describe('mount', () => {
    it('should transform labels', async () => {
      expect(await compile('mount: { console.log(count); }')).toMatchSnapshot();
    });
    it('should transform labels with expressions', async () => {
      expect(await compile('mount: myFunction;')).toMatchSnapshot();
    });
  });
  describe('cleanup', () => {
    it('should transform labels', async () => {
      expect(await compile('cleanup: { console.log(count); }')).toMatchSnapshot();
    });
    it('should transform labels with expressions', async () => {
      expect(await compile('cleanup: myFunction;')).toMatchSnapshot();
    });
  });
  describe('error', () => {
    it('should transform labels', async () => {
      expect(await compile('error: { console.log(count); }')).toMatchSnapshot();
    });
    it('should transform labels with expressions', async () => {
      expect(await compile('error: myFunction;')).toMatchSnapshot();
    });
  });
});
