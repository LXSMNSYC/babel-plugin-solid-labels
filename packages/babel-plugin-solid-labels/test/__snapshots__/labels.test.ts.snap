// Vitest Snapshot v1

exports[`labels > $ > should transform labels with block expressions 1`] = `
"import { createEffect as _createEffect } from \\"solid-js\\";

_createEffect(() => {
  console.log(x);
});"
`;

exports[`labels > $ > should transform labels with expressions 1`] = `
"import { createEffect as _createEffect } from \\"solid-js\\";

_createEffect(() => console.log(x));"
`;

exports[`labels > $ > should transform labels with variable declarations 1`] = `
"import { createMemo as _createMemo } from \\"solid-js\\";

const _x = _createMemo(() => 0);"
`;

exports[`labels > cleanup > should transform labels 1`] = `
"import { onCleanup as _onCleanup } from \\"solid-js\\";

_onCleanup(() => {
  console.log(count);
});"
`;

exports[`labels > cleanup > should transform labels with expressions 1`] = `
"import { onCleanup as _onCleanup } from \\"solid-js\\";

_onCleanup(myFunction);"
`;

exports[`labels > computed > should transform labels 1`] = `
"import { createComputed as _createComputed } from \\"solid-js\\";

_createComputed(() => {
  console.log(count);
});"
`;

exports[`labels > computed > should transform labels with expressions 1`] = `
"import { createComputed as _createComputed } from \\"solid-js\\";

_createComputed(myFunction);"
`;

exports[`labels > computed > should transform labels with names 1`] = `
"import { createComputed as _createComputed } from \\"solid-js\\";

_createComputed(() => {
  console.log(count);
}, undefined, {
  name: \\"countLog\\"
});"
`;

exports[`labels > effect > should transform labels 1`] = `
"import { createEffect as _createEffect } from \\"solid-js\\";

_createEffect(() => {
  console.log(count);
});"
`;

exports[`labels > effect > should transform labels with expressions 1`] = `
"import { createEffect as _createEffect } from \\"solid-js\\";

_createEffect(myFunction);"
`;

exports[`labels > effect > should transform labels with names 1`] = `
"import { createEffect as _createEffect } from \\"solid-js\\";

_createEffect(() => {
  console.log(count);
}, undefined, {
  name: \\"countLog\\"
});"
`;

exports[`labels > error > should transform labels 1`] = `
"import { onError as _onError } from \\"solid-js\\";

_onError(() => {
  console.log(count);
});"
`;

exports[`labels > error > should transform labels with expressions 1`] = `
"import { onError as _onError } from \\"solid-js\\";

_onError(myFunction);"
`;

exports[`labels > memo > should transform labelled variables 1`] = `
"import { createMemo as _createMemo } from \\"solid-js\\";

const _count = _createMemo(() => 0);"
`;

exports[`labels > memo > should transform labelled variables and add name 1`] = `
"import { createMemo as _createMemo } from \\"solid-js\\";

const _count = _createMemo(() => 0, undefined, {
  name: \\"count\\"
});"
`;

exports[`labels > memo > should variable read 1`] = `
"import { createMemo as _createMemo } from \\"solid-js\\";

const _count = _createMemo(() => 0);

console.log(_count());"
`;

exports[`labels > mount > should transform labels 1`] = `
"import { onMount as _onMount } from \\"solid-js\\";

_onMount(() => {
  console.log(count);
});"
`;

exports[`labels > mount > should transform labels with expressions 1`] = `
"import { onMount as _onMount } from \\"solid-js\\";

_onMount(myFunction);"
`;

exports[`labels > renderEffect > should transform labels 1`] = `
"import { createRenderEffect as _createRenderEffect } from \\"solid-js\\";

_createRenderEffect(() => {
  console.log(count);
});"
`;

exports[`labels > renderEffect > should transform labels with expressions 1`] = `
"import { createRenderEffect as _createRenderEffect } from \\"solid-js\\";

_createRenderEffect(myFunction);"
`;

exports[`labels > renderEffect > should transform labels with names 1`] = `
"import { createRenderEffect as _createRenderEffect } from \\"solid-js\\";

_createRenderEffect(() => {
  console.log(count);
}, undefined, {
  name: \\"countLog\\"
});"
`;

exports[`labels > signal > should transform labelled variables 1`] = `
"import { createSignal as _createSignal } from \\"solid-js\\";

const [_count, _setcount] = _createSignal(0);"
`;

exports[`labels > signal > should transform labelled variables and add name 1`] = `
"import { createSignal as _createSignal } from \\"solid-js\\";

const [_count, _setcount] = _createSignal(0, {
  name: \\"count\\"
});"
`;

exports[`labels > signal > should variable read 1`] = `
"import { createSignal as _createSignal } from \\"solid-js\\";

const [_count, _setcount] = _createSignal(0);

console.log(_count());"
`;

exports[`labels > signal > should variable write 1`] = `
"import { createSignal as _createSignal } from \\"solid-js\\";

const [_count, _setcount] = _createSignal(0);

_setcount(() => 1);"
`;
